calling-apis/managing-asynchronous-calls.md

```diff
@@ -0,0 +1,277 @@
+    }
+  })
+  .then( (response) => {
+    latitude = response.data[0].lat;
+    longitude = response.data[0].lon;
+    console.log('success in findLatitudeAndLongitude!', latitude, longitude);
+  })
+  .catch( (error) => {
+    console.log('error in findLatitudeAndLongitude!');
+  });
+
+  return {
+      seattleLat: latitude,
+      seattleLon: longitude
+  }
+}
+
+const findLocation = (latitude, longitude) => {
+  axios.get('https://us1.locationiq.com/v1/reverse.php',
+  {
+    params: {
+      key: YOUR_API_KEY,
+      format: 'json',
+      lat: latitude,
+      lon: longitude
+    }
+  })
+  .then( (response) => {
+    console.log('success in findLocation!', response.data);
+    return response.data;
+  })
+  .catch( (error) => {
+    console.log('error in findLocation!');
+  });
+}
+
+const seattleCoordinates = findLatitudeAndLongitude('Seattle, Washington, USA');
+
+const locations = findLocation(seattleCoordinates.seattleLat, seattleCoordinates.seattleLon);
+
+console.log(locations);
+```
+<!-- prettier-ignore-end -->
+
+This code:
+
+- Defines `findLatitudeAndLongitude`, which searches for latitude and longitude with the given query
+- Defines `findLocation`, which searches for location with the given latitude and longitude
+- Finds the coordinates of Seattle by calling `findLatitudeAndLongitude` and puts it in `seattleCoordinates`
+- Finds the location results by calling `findLocations` and puts it in `locations`
+
+However, when we run this code, our console outputs this:
+
+```
+error in findLocation!
+success in findLatitudeAndLongitude! 47.6038321 -122.3300624
+```
+
+Raffy's experiment doesn't look successful at all! What's noticeable is that `error in findLocation!` prints out **_before_** `success in findLatitudeAndLongitude!` does.
+
+It's apparent that the API call in `findLatitudeAndLongitude` works. We can get back a latitude and longitude just fine!
+
+Why would there be an error when calling `findLocation`? And why would this print **_before_** we see a print statement from `findLatitudeAndLongitude`?
+
+<!-- Question 1 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: paragraph
+* id: e723527e
+* title: Managing Asynchronous Calls
+##### !question
+
+Make a hypothesis:
+
+Why would there be an error when calling `findLocation`? And why would this print **_before_** we see a print statement from `findLatitudeAndLongitude`?
+
+##### !end-question
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+### Debugging Our Problem
+
+After debugging, Raffy should hone in on one specific line:
+
+```js
+const locations = findLocation(seattleCoordinates.seattleLat, seattleCoordinates.seattleLon);
+```
+
+Raffy found these two facts while debugging:
+
+- `seattleCoordinates.seattleLat` _and_ `seattleCoordinates.seattleLon` are `undefined` at the time of calling `findLocation`
+- The print statements in `findLocation` printed, meaning that `findLocation` was definitely called and got into the `catch` block
+
```

Comment:
I think it would be helpful to include more details about how he continues his investigation, including a snippet of the additional logging he adds.
The next comment has the suggested markdown for this.
No suggestion because of the code block
(there should be blank lines before and after the new content, but github seems to strip them from the comments)

Reply:
Raffy decides to trace further back to check the values in the object being assigned to `seattleCoordinates`. He knows this is the return value from `findLatitudeAndLongitude`, so he adds some logging before the return statement to check on the `latitude` and `longitude`.

<!-- prettier-ignore-start -->
```js
  console.log('latitude:', latitude);
  console.log('longitude:', longitude);

  // existing return in findLatitudeAndLongitude
  return {
      seattleLat: latitude,
      seattleLon: longitude
  }
```
<!-- prettier-ignore-end -->

He is surprised by the result.

When the object being returned from `findLatitudeAndLongitude` is created, `latitude` and `longitude` are both `undefined`. On top of that, the logging statements print out before the success message prints out. The return is running before the API call returns!